Add print options to allow digging, filtering and formatting

ronkorving opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

Closes https://github.com/serverless/plugins/issues/164

This puts the print plugin a bit on steroids. It gives it three options:

```
serverless print --help
Plugin: Print
print ......................... Print your compiled and resolved config file
    --format ........................... Print configuration in given format ("yaml", "json", "text"). Default: yaml
    --path ............................. Optional path that lets you print only a sub-object (eg: "provider.name")
    --filter ........................... Optional filter to apply to the configuration object ("keys")
```

Example to extract all lambda functions from a configuration:

```sh
FUNCTIONS=`serverless print --format text --path functions --filter keys`
echo $FUNCTIONS
```

## How did you implement it:

I added the options to the print plugin, and implemented their logic in the print() method. All are unit tested.

## How can we verify it:

Run `serverless print --help` and apply it to any configuration. You can easily switch the output between yaml (default), json and plain text using `--format`. For `--path`, write a dot-separated string to dig into the configuration structure and return only a sub-object. Use `--filter keys` to return not the object, but its keys.

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [ ] Write documentation
- [ ] Fix linting errors
- [ ] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

horike37 commented on lib/plugins/print/print.js:
```diff
@@ -91,9 +104,44 @@ class Print {
+            let out;
+
+            if (this.options.format === 'text') {
+              if (Array.isArray(conf)) {
```
Please use `lodash` for consistency of other places since we have decided to use it for if-check.

ronkorving commented on lib/plugins/print/print.js:
```diff
@@ -91,9 +104,44 @@ class Print {
+              if (this.options.filter === 'keys') {
+                conf = Object.keys(conf);
+              } else {
+                throw new this.serverless.classes.Error('Filter must be "keys"');
```
It's perfectly valid to throw inside a promise. It will reject the promise, whereas `Promise.reject()` will return a *new* promise. Having said that, I'm all for consistency, so if you insist I'll change it.

ronkorving commented on lib/plugins/print/print.js:
```diff
@@ -91,9 +104,44 @@ class Print {
+            // print configuration in the specified format
+            let out;
+
+            if (this.options.format === 'text') {
```
Yes, it's plain text, and currently only works for scalar values (strings, booleans, numbers) and arrays of scalar values. Do you have any suggestions how it should output objects? My purpose with outputting text was so that it would be easily consumable by shell scripts (which JSON and YAML aren't).

horike37 replied:
OK. How about adding only works for scalar values to the documentation? Otherwise, users may misunderstand to get any errors.

ronkorving replied:
Understood, I will do that 👍

ronkorving commented on lib/plugins/print/print.js:
```diff
@@ -91,9 +104,44 @@ class Print {
+            // dig into the object
+            if (this.options.path) {
+              const steps = this.options.path.split('.');
+              for (const step of steps) {
```
I'm traversing an object in a synchronous way, I'm a bit confused why you would want to make that operation behave asynchronously? Are you saying this project does not allow any for-loops inside of functions that happen to return a promise?

ronkorving replied:
For now, I will change it to be `steps.forEach()`, which is more consistent with the rest of the codebase.

horike37 replied:
Thank you for response 👍 
Yes, I know there are a lot of synchronous loops in promises in our codebase.
However, we(the framework maintainer team) have decided to use asynchronous way before as possible as we can, because current unit tests have sometimes failed with an unknown cause, and we are considering the cause is that haven't caught errors properly because of mixing synchronous/asynchronous. 

What do you think?

ronkorving replied:
Maybe you can show me what kind of problems you have seen? Maybe I can help.

horike37 replied:
@ronkorving 
Thank you for validating the behavior 👍 

>Maybe you can show me what kind of problems you have seen? 

Unit tests of the framework have been flaky for a long time, i.e. it fails sometimes with uncaught or timeout errors but it fixes with re-building Travis again in many cases. 
https://travis-ci.org/serverless/serverless/jobs/312633967
https://travis-ci.org/serverless/serverless/builds/393374487

We have found some place which don't handle promise error, and maybe this leads the flaky tests, but it is hard to remove it all places.
Therefore we recently unify it with BbPromse as possible as we can.

ronkorving replied:
The timeout seems to be set to 2 seconds, and if I/O is slow (which it can be, especially on a free service like TravisCI), maybe it's just a matter of increasing the timeout value to something a bit higher? The tests that fail in the job you linked to are the exact same 2 tests in Node 4 and 6. Is it always those 2 that time out?

https://mochajs.org/#timeouts

ronkorving commented:
@horike37 Thank you for the review! I've left some questions I hope you can answer. I will update some code to be more in-line with the existing code style (error handling, lodash).

ronkorving commented:
@horike37 I've added documentation to the 4 files you mentioned. Let me know if it's to your liking. I have also added a few more error conditions, and renamed `filter` to `transform`, because I think it matches the operation better.

ronkorving commented:
Please let me know if this PR needs anything else for it to be ready to be merged. Thanks so far 👍

ronkorving commented:
@horike37 Thanks!
